### DISCUSSION TITLE
请问“常用”栏的三个选项和“编码器”的音频选项有什么区别？

### DISCUSSION THREAD

**Discussion by MC-dusk** [zyzsdy/NegativeEncoder#69] [Q&A] [answered]
如图：
![image](https://github.com/zyzsdy/NegativeEncoder/assets/52311860/7514f0b2-b0d7-4ab8-8362-6dc943bcb8c7)
![image](https://github.com/zyzsdy/NegativeEncoder/assets/52311860/c22522c2-9691-43ae-85d6-79a3eefde798)
感觉功能重复了

**Comment by zyzsdy** [comment, marked as answer]
ffmpeg pipe这个功能是为了使用ffmpeg取代NVENC的硬件解码器。并且在使用第三个按钮“处理音频与视频”时，会由ffmpeg对音频实行变换。在使用ffmpeg pipe时，首页的音频选项实际上被绕过，发送给nvencc编码器的始终为无音频流模式。

而在首页的音频选项则是为了设置nvencc编码器自带的音频处理。

接下来你可能会关心这个问题：当你使用nvnecc的软件解码时实际上也是使用了ffmpeg进行解码，这和使用ffmpeg pipe有何不同？
简单的说：我们发现了一些特殊的容器格式，它们通常是由一些直播网站提供的串流（stream），其帧率是动态的。对于这些动态帧率视频（VFR），nvencc提供了一些功能可以修复它，重新对齐音频和视频，包括读取并重新写入时间码，或是强制将视频转为CFR，这些功能被放置在首页的“高级”-“音频同步”选项中。

然而并不是所有的音频同步问题都能由这种方式修复，有些平台提供的原始视频流十分特殊，即使是nvencc的forcecfr模式也无法修复。因此我们选择由独立的ffmpeg将视频变成yuv420格式的一帧帧图像，此时由ffmpeg决定复制一部分帧，已经变为CFR，再通过管道提供给nvencc编码成视频文件，这就是这里的ffmpeg pipe功能。

对于无需处理音频或是稍后手动处理的音频的功能，需要使用“复制音频”“无音频”按钮。然而对于一部分VFR视频，其音频采样率可能也是动态的，因此我们通过ffmpeg同时处理音频和视频的问题，将它们都转换成固定帧率/采样率的。对于这种情况适用于“处理音频与视频按钮”。

NegativeEncoder最早是为了字幕组需求而开发，因此各种功能都适用于字幕组处理片源的习惯。为VFR视频添加字幕，尤其是带有特效的字幕非常困难，因此为了音频同步，都是转换为CFR在处理的。因此有了这些看似重复的功能。

开发团队的几个人都没有日常检查Discussion的习惯，抱歉回复晚了。

**Reply by MC-dusk** [reply]
谢谢，意思就是第一页“编码器”页，解码模式为硬件解码对应视频音频硬解硬编，解码模式为软件解码对应视频音频软解硬编依靠nvencc修复同步，“常用”页的ffmpeg pipe对应视频音频软解硬编依靠ffmpeg修复同步。
如果ffmpeg pipe选择“处理音频与视频”，音频码率就无法使用首页设置的值，而是使用默认值128k。
